adam/learner/pursuit.py

```diff
@@ -370,68 +401,50 @@ def learning_step(
         return hypothesis_is_confirmed
 
     @attrs(frozen=True)
-    class PartialMatch:
+    class PartialMatch(Generic[HypothesisT2]):
         """
+        A class to hold hypothesis match information, such as the partial_match_hypothesis.
         *match_ratio* should be 1.0 exactly for a perfect match.
         """
 
-        matching_subgraph: Optional[PerceptionGraphPattern] = attrib(
-            validator=optional(instance_of(PerceptionGraphPattern))
-        )
         num_nodes_matched: int = attrib(validator=instance_of(int), kw_only=True)
         num_nodes_in_pattern: int = attrib(validator=instance_of(int), kw_only=True)
 
-        def matched_exactly(self) -> bool:
-            return self.num_nodes_matched == self.num_nodes_in_pattern
-
-        def match_ratio(self) -> float:
-            return self.num_nodes_matched / self.num_nodes_in_pattern
+        @abstractmethod
+        def partial_match_hypothesis(self) -> Optional[HypothesisT2]:
+            """
+            Returns a hypothesis corresponding to the partial match.
+            """
 
-    def _compute_match_ratio(
-        self, pattern: PerceptionGraphPattern, graph: PerceptionGraph
-    ) -> "PursuitLanguageLearner.PartialMatch":
-        """
-        Computes the fraction of pattern graph nodes of *pattern* which match *graph*.
-        """
-        hypothesis_pattern_common_subgraph = get_largest_matching_pattern(
-            pattern,
-            graph,
-            debug_callback=self._debug_callback,
-            graph_logger=self._graph_logger,
-            ontology=self._ontology,
-            matching_objects=True,
-        )
-        self.debug_counter += 1
+        @abstractmethod
+        def matched_exactly(self) -> bool:
+            """
+            Returns a boolean indicating whether the matching was an exact match
+            """
 
-        leading_hypothesis_num_nodes = len(pattern)
-        num_nodes_matched = (
-            len(hypothesis_pattern_common_subgraph.copy_as_digraph().nodes)
-            if hypothesis_pattern_common_subgraph
-            else 0
-        )
-        return PursuitLanguageLearner.PartialMatch(
-            hypothesis_pattern_common_subgraph,
-            num_nodes_matched=num_nodes_matched,
-            num_nodes_in_pattern=leading_hypothesis_num_nodes,
-        )
+        @abstractmethod
+        def match_score(self) -> float:
+            """
+            Returns a score (e.g. ratio of matching nodes) indicating how much the hypotheses match
```

Comment:
This is just some score in [0.0, 1.0] where 0.0 indicates no match at all and 1.0 indicates a perfect match.

For the general interface we don't put any further requirements on the semantics.